parlai/scripts/train_model.py

```diff
@@ -251,25 +257,43 @@ def run_eval(valid_world, opt, datatype, max_exs=-1, write_log=False):
     :param int max_exs:
         limits the number of examples if max_exs > 0
     """
-    if valid_world is None:
+    if valid_worlds is None:
         # This isn't the primary worker, so we can just skip evaluation
         return None
 
     print('[ running eval: ' + datatype + ' ]')
-    valid_world.reset()
-    cnt = 0
-    while not valid_world.epoch_done():
-        valid_world.parley()
-        if cnt == 0 and opt['display_examples']:
-            print(valid_world.display() + '\n~~')
-            print(valid_world.report())
-        cnt += valid_world.opt['batchsize']
-        if max_exs > 0 and cnt >= max_exs:
-            # note this max_exs is approximate--some batches won't always be
-            # full depending on the structure of the data
-            break
-    valid_report = valid_world.report()
-    valid_world.reset()  # this makes sure agent doesn't remember valid data
+    reports = []
+    for v_world in valid_worlds:
```

Comment:
If we kept as a MultiWorld and split this into a recursive function we could address Jason’s concerns from the meeting while also not changing the type of the original variable (making it simpler diff)

Reply:
sorry i don't understand how that would fix it if we kept as a multiworld?